### PR TITLE
WIP: open fd passing alongside socket

### PR DESCRIPTION
**Pull request by baloo**
See https://github.com/tokio-rs/tokio/issues/570

I spent some time in strace to get it right, and it looks correct.

Missing:
 - [ ] Figure out the fd handling (when to take ownership, when to close an open fd) ...
 - [ ] Fix failing test on travis (tests are passing locally)

### CODE REVIEW

File: src/cmsg.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            if remaining.len() < hdr_len {
+                panic!("not enough bytes for length");
+            }
+
+            let mut cur = Cursor::new(remaining);
+
+            #[cfg(target_pointer_width="32")]
+            let len = cur.read_u32::<NativeEndian>().unwrap() as usize;
+            #[cfg(target_pointer_width="64")]
+            let len = cur.read_u64::<NativeEndian>().unwrap() as usize;
```
Comment (Ralith) [inline src/cmsg.rs:152]: Perhaps `cur.read_uint::<NativeEndian>(mem::size_of::<*const ()>())`?

**Comment by Ralith** [comment]
It would be cool to extract some of this out into a reusable general-purpose cmsg-parsing lib.

**Comment by baloo** [comment]
Tests are passing now. I think I got the memory management right with cmsg (which was my biggest issue with it).
@Ralith I'd be happy to do that, I would probably need some help with API (as I'm not sure I would get it correct for all use-cases).

**Comment by baloo** [comment]
turns out the failing tests on travis are on macos, and the `libc::msghdr` there has different field types.

**Comment by baloo** [comment]
turns out the failing tests on travis are on macos, and the `libc::msghdr` there has different field types, i need to rework this a bit.

**Comment by baloo** [comment]
@carllerche tokio-uds being the biggest consumer of this crate, I'd love to have your feedback on the API I'm adding here. Thank you!

cc @tobz

**Comment by Ralith** [comment]
It looks like libc defines `CMSG_FIRSTHDR`, `CMSG_NXTHDR`, etc. on linux, though not yet on BSD. Would it make sense to use those rather than reimplementing their logic?

**Comment by baloo** [comment]
@Ralith I missed those in libc (I actually haven't looked it up, convinced no C macros was exported to rust, looks like I missed something). 
I don't have a BSD version of libc right now. 

how about importing this for other unix systems?
https://github.com/rust-lang/libc/blob/728f6e06b7c3b256623340fe39c49fab23c923c4/src/unix/notbsd/mod.rs#L979

**Comment by Ralith** [comment]
`man cmsg` suggests that it's a POSIX standard, so you may be able to find it on other unixes, and therefore potentially get a PR accepted to make those definitions more broadly exposed by libc.

**Comment by Ralith** [comment]
On inspection, every Unixlike I've checked (FreeBSD, OpenBSD, macOS, Linux, android) defines those macros. Annoyingly, all the definitions look slightly different. Still, we should be able to get them into libc.

**Comment by pfmooney** [comment]
If you're still interested in support for this, I would encourage you to work with the `mio` folks to get it integrated into the native UDS support present in the 0.7.x series.

**Pull request closed without merging**